feat(linter): add support for nested config files

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/7408

This is the initial implementation for nested config support, without the `extends` keyword factoring into it yet. So we only load a single config file for any given file still, but it will always be the one closest to the file being linted.

Performance is somewhat of a big TODO here, obviously there is some inherent cost to adding support for nested configurations since we need to check if any directory contains an `.oxlintrc.json` file, load it if it exists, and then for every file below that directory check if we should use it. I've left a few TODOs of different ways we might be able to cache more data at the cost of memory but at the potential gain of some CPU time. Optimizing iteration over files/directories is one way in which we might be able to make this faster, so we never need to visit the same directory twice.

For now, this will remain an experimental feature that should not affect the current functionality until we enable it for everyone. PRs stacked on top of this will focus on ensuring that other features work well with this, such as `--config`, `-D <rule>`, and then we'll add new features like `extends`.

The configuration loading all happens upfront currently, while the config resolution happens at lint-time. There is a new `nested_configs` field on the `Linter` that passes all of the configs that are available, so it can pick which one to use later. Eventually we'll refactor this to not have two duplicate config fields, but I'm leaving it as-is to ensure we're not messing up the existing functionality.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9153?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9153** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9153?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9153?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9152** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9152?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9153).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented on crates/oxc_linter/src/lib.rs:
```diff
@@ -62,11 +67,23 @@ pub struct Linter {
     config: ConfigStore,
+    // TODO(refactor): remove duplication with `config` field when nested config is
+    // standardized, as we do not need to pass both at that point
+    nested_configs: FxHashMap<PathBuf, ConfigStore>,
```
Can this somehow be later marked as non empty? 
Currently this works because `self.config` is then used. 
When no configuration file is found, the default configuration should be used.

camchenry replied:
This is a good question, I'm actually less certain now that we'll need to get rid of the `config` field here, as we do still care about the config in the current working directory or the default config because it's what we use for including/excluding files, for example. I'll think on this a bit more as I'm working on this

Sysix commented:
Thanks for the great work ❤️  I added some comments and I have a question:

> ESLint then searches up the directory structure, merging any .eslintrc files it finds along the way until reaching either an .eslintrc file with root: true or the root directory.

_https://eslint.org/docs/latest/use/configure/configuration-files-deprecated#cascading-and-hierarchy_

Is this part missing? I could not find the code where you merge the different configurations like `overrides`.

camchenry commented on apps/oxlint/src/lint.rs:
```diff
@@ -168,6 +170,56 @@ impl Runner for LintRunner {
+            let mut directories = FxHashSet::default();
+            for path in &paths {
+                if let Some(directory) = path.parent() {
+                    // TODO(perf): benchmark whether or not it is worth it to produce the config files here without
```
I addressed this a little bit in the comment here: https://github.com/oxc-project/oxc/issues/7408#issuecomment-2513242536

> In ESLint, configuration files are resolved by using the file in the current directory or nearest parent directory as the base configuration file, and then subsequently merging that with each file in each parent directory, up to the root directory. The base configuration file takes precedence over any parent configurations.
> 
> However, this can create situations where you have more rules enabled than you want, such as rules that shouldn't apply for test files, or rules for packages you aren't using in a subdirectory. ESLint solves this by supporting adding a root: true property to the configuration file, which indicates that it should not find and merge configuration files in parent directories. You can still extend other configuration files though.
>
> In oxlint, we will not automatically search upwards in the file hierarchy for configuration files. We will only use the closest configuration file to the file currently being linted. Every configuration file will essentially be treated as if it already contained root: true. This makes configuration resolution faster, because we don't need to do any further searching once we've find the closest configuration file, and chances are that we've already parsed it and cached it, since it's in a parent directory. This also makes it easier for developers to understand how the final linting configuration is formed, because you can always just find the

We've diverged slightly from the ESLint implementation. Based on my experience, I think `root: true` is a little bit of a confusing feature and also means we would end up doing more config merging than is maybe necessary. This way, we require users to explicitly say which configs they want to be merged.

I think you're right though, since every config file basically has `root: true`, we should be able to stop iterating as soon as we find a config file, as that is the nearest one.

camchenry commented:
@Boshen I'll address the merge conflicts and then merge this

camchenry commented:
I'm going to use this repository as a candidate for benchmarking: https://github.com/microsoft/fluentui, as it has a lot of nested `.eslintrc` files which should make it a good example. There are 14,836 lintable files in the repository.

Sysix commented:
> In the worst case, it looks like it is only up to 6% slower, and only about 2% slower on average, which is pretty good for a repo of this size!

I guess this is only when no `.oxlintrc.json` file is found.
The result would be interesting when we set `.eslintrc.json` as the default for this repo. 

Could you please resolve the conflicts? The changes are good for me now 👍 

A future idea: 
Skip search-iteration of paths-(files) with the same directory. The result should be known already.

camchenry commented on apps/oxlint/src/lint.rs:
```diff
@@ -168,6 +170,56 @@ impl Runner for LintRunner {
+                }
+            }
+            for directory in directories {
+                // TODO: how should we handle a nested config error?
```
I think this should be handled correctly for now, we throw an error below when building the `ConfigStore`. More errors will be thrown later when merging configs too 😄

camchenry commented:
@Sysix Sorry, you caught me in the middle of benchmarking and I left the comment in a half written state for a few hours 😅 I updated the post with a better benchmark. 

Overall, I think performance is looking pretty good (so far), in the range of 5-6% slower so far, so not worth optimizing quite yet.

camchenry committed:
feat(linter): add support for nested config files (#9153)

- part of https://github.com/oxc-project/oxc/issues/7408

This is the initial implementation for nested config support, without the `extends` keyword factoring into it yet. So we only load a single config file for any given file still, but it will always be the one closest to the file being linted.

Performance is somewhat of a big TODO here, obviously there is some inherent cost to adding support for nested configurations since we need to check if any directory contains an `.oxlintrc.json` file, load it if it exists, and then for every file below that directory check if we should use it. I've left a few TODOs of different ways we might be able to cache more data at the cost of memory but at the potential gain of some CPU time. Optimizing iteration over files/directories is one way in which we might be able to make this faster, so we never need to visit the same directory twice.

For now, this will remain an experimental feature that should not affect the current functionality until we enable it for everyone. PRs stacked on top of this will focus on ensuring that other features work well with this, such as `--config`, `-D <rule>`, and then we'll add new features like `extends`.

The configuration loading all happens upfront currently, while the config resolution happens at lint-time. There is a new `nested_configs` field on the `Linter` that passes all of the configs that are available, so it can pick which one to use later. Eventually we'll refactor this to not have two duplicate config fields, but I'm leaving it as-is to ensure we're not messing up the existing functionality.
